### PR TITLE
Adding support for deterministic GUID

### PR DESCRIPTION
**Pull request by gled4er**
Hello @cgillum,

May I ask you to review the code in this PR that is addressing the issue for generating deterministic `Guid` from the `DurableOrchestrationContext` as described here https://github.com/Azure/azure-functions-durable-extension/issues/377.

Thank you!

### CODE REVIEW

**Comment by SimonLuckenuik** [comment]
Hi @gled4er,

I just had a quick look at your code and I believe that the following behavior is not what I would expect as a user of the API:
```
            Guid firstGuid = ctx.NewGuid();
            string output = await ctx.CallActivityAsync<string>(nameof(TestActivities.Hello), input);
            Guid secondGuid = ctx.NewGuid();
            return firstGuid == secondGuid;
```
And in your test you check for `Assert.Equal("True", status.Output.ToString());`. 
I would expect firstGuid and secondGuid to be different, not equal.

The goal of NewGuid would be to enable us to create multiple GUIDs prior passing them to different activities. For example, I could create 10 GUIDs, do a fan-out and provide a different GUID to each activity part of the fan-out. The name also implies that it's creating a new GUID at each call, like `Guid.NewGuid()`, which is not the case in your unit test.

Simon

**Comment by gled4er** [comment]
Hello Simon,

Thank you for the great feedback!

Will you mind passing a name that,can be any string to the NewGuid method? If this is possible the behavior will be fixed easily. I can make the name optional parameter but when provided will generate different Guid-s per different names.

Let me know if this is acceptable for you.

Thank you!

**Comment by cgillum** [comment]
Just like `Guid.NewGuid()` doesn't take a parameter, `ctx.NewGuid()` also shouldn't have a parameter.

Would it be possible to simply include a monotonically incrementing counter variable as part of the name? It seems you could increment this counter variable every time `NewGuid` gets called.

**Comment by gled4er** [comment]
Hello @cgillum,

Great idea!

I will modify the code.

Thank you!

**Comment by gled4er** [comment]
Hello @SimonLuckenuik  @cgillum,

I modified the code according our discussion.

May I ask you to take a look and let me know what should be fixed next?

Thank you!

**Comment by gled4er** [comment]
Hello @SimonLuckenuik,

I was wondering if you had time to review the changes after your last comment.

I will be happy to hear if we are getting closer to support your scenario.

Thank you!

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs [hunk 1]
```diff
@@ -100,6 +102,15 @@ public override T GetInput<T>()
             return MessagePayloadDataConverter.Default.Deserialize<T>(this.serializedInput);
         }
 
+        /// <inheritdoc />
+        public override Guid NewGuid()
+        {
+            this.newGuidCounter++;
+            StringBuilder newGuidNameStringBuilder = new StringBuilder(this.InstanceId);
+            newGuidNameStringBuilder.Append(this.newGuidCounter);
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs]: Using a `StringBuilder` to concatenate two strings is actually more overhead than doing string concatenation directly. I suggest simply:
```csharp
string guidNameValue = this.InstanceId + this.newGuidCounter.ToString();
this.newGuidCounter++;
```

File: test/Common/TestOrchestrations.cs [hunk 2]
```diff
@@ -28,6 +28,16 @@ public static async Task<string> SayHelloWithActivity([OrchestrationTrigger] Dur
...
         }
 
+        public static async Task<bool> SayHelloWithActivityWithDeterministicGuid([OrchestrationTrigger] DurableOrchestrationContext ctx)
+        {
+            string input = ctx.GetInput<string>();
+            Guid firstGuid = ctx.NewGuid();
+            Guid secondGuid = ctx.NewGuid();
+            string output = await ctx.CallActivityAsync<string>(nameof(TestActivities.Hello), input);
+            Guid thirdGuid = ctx.NewGuid();
+            return firstGuid != secondGuid && firstGuid != thirdGuid && secondGuid != thirdGuid;
```
Comment (cgillum) [inline test/Common/TestOrchestrations.cs]: This is a good basic test. Consider two more possible tests:
1. A test that ensures if you create 10K+ GUIDs in your orchestration they are all unique.
2. A test that ensures the same GUID values are generated during replays.

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs (hunk 1 above)
Comment (gled4er) [reply src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs]: Hello @cgillum,

Great!

I will modify it!

Thank you!

File: test/Common/TestOrchestrations.cs (hunk 2 above)
Comment (gled4er) [reply test/Common/TestOrchestrations.cs]: Hello @cgillum,

Thank you for the great feedback.

I will add the tests.

Thank you!

**Comment by gled4er** [comment]
Hello @cgillum,

I worked on the requested changes during the last review iteration.

May I ask you to take another look when you have time?

Thank you!

File: test/Common/TestOrchestrations.cs [hunk 3]
```diff
@@ -28,6 +28,55 @@ public static async Task<string> SayHelloWithActivity([OrchestrationTrigger] Dur
...
+        }
+
+        public static bool VerifyUniqueGuids([OrchestrationTrigger] DurableOrchestrationContext ctx)
+        {
+            List<Guid> guids = new List<Guid>(10000);
```
Comment (cgillum) [inline test/Common/TestOrchestrations.cs]: A `HashSet<Guid>` would be much more natural and efficient than `List<Guid>` for a test like this.

**Comment by gled4er** [comment]
Hello @cgillum,

Thank you very much for the support on this PR!

Thank you!

File: test/Common/TestOrchestrations.cs (hunk 3 above)
Comment (gled4er) [reply test/Common/TestOrchestrations.cs]: Great!

I will change it!

Thank you!

**Pull request merged**